anvil/src/xwayland/mod.rs

```diff
@@ -193,6 +193,21 @@ impl X11State {
                     // wayland socket). Thus, we could receive these two in any order. Hence, it
                     // can happen that we get None below when X11 was faster than Wayland.
 
+                    let location = {
+                        match self.conn.get_geometry(msg.window)?.reply() {
+                            Ok(geo) => (geo.x.into(), geo.y.into()),
+                            Err(err) => {
+                                error!(
+                                    self.log,
+                                    "Failed to get geometry for {:x}, perhaps the window was already destroyed?",
+                                    msg.window;
+                                    "err" => format!("{:?}", err),
+                                );
+                                (0, 0)
+                            }
+                        }
+                    };
+
```

Comment:
I would rather like to request this info in `new_window` instead and removing the need to store the location in `unpaired_surfaces`. That should be possible as well, right?

Reply:
Yes and no. To avoid a deadlock with the X11 server, whenever you write to the X11 socket, you also have to read data. x11rb does that. Thus, even just sending a request can end up reading from the X11 socket. That's bad, since it could now happen that an X11 event was read from the socket and is buffered in x11rb, but the socket is no longer readable. Thus, the calloop integration that I add in this PR would not wake up and the event would sit in the buffer until the next event comes.

For this reason, some more magic is necessary before anything can interact with the X11 connection outside of the event handling (but such magic will definitely be necessary in the future). My best idea on how to tackle this so far is something like calloop's `make_ping()`: Whenever we do anything with the X11 connection, we also send a ping. This then allows us to later check if some more X11 events are pending.

Also, your suggestion would make `new_window()` fallible / have to return a `Result`, which means I have to think even more about error handling. Currently I managed to hide everything behind few `unwrap()`s. ;-)